### PR TITLE
[strings] Fix split and rsplit implementations

### LINKED ISSUE #23616: [jit] string.split only splits on newlines
It should split on any whitespace, right now it just splits on `' '`

```python
def fn(x):
    # type: (str)
    return x.split()

s = "a b\nc\td"
print(fn(s))
print(torch.jit.script(fn)(s))
```
outputs

```
['a', 'b', 'c', 'd']
['a', 'b\nc\td']
```

cc @suo

### PR DESCRIPTION
**Pull request by bwasti**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#23669 [strings] Fix split and rsplit implementations**

Differential Revision: [D16603705](https://our.internmc.facebook.com/intern/diff/D16603705)

Fixes #23616

### CODE REVIEW

**Comment by bwasti** [comment]
fixes #23616

File: test/test_jit_string.py [hunk 1]
```diff
@@ -263,9 +264,10 @@ def test_rpartition():
         self.checkScript(test_rpartition, ())
 
         def test_split():
-            # type: () -> Tuple[List[str], List[str], List[str], List[str], List[str], List[str], List[str], List[str], List[str]]
+            # type: () -> Tuple[List[str], List[str], List[str], List[str], List[str], List[str], List[str], List[str], List[str], List[str]]
```
Comment (driazati) [inline test/test_jit_string.py]: Why not make this `List[List[str]]` so we don't have to keep editing it? Or even just delete the type comment since return types can be inferred

File: test/test_jit_string.py [hunk 2]
```diff
@@ -1,5 +1,6 @@
 from test_jit import JitTestCase
-
+import sys
+sys.setrecursionlimit(10000)
```
Comment (driazati) [inline test/test_jit_string.py]: Why is this needed?

File: torch/csrc/jit/register_string_ops.cpp [hunk 3]
```diff
@@ -547,48 +547,94 @@ auto reg_str_ops_2 = torch::RegisterOperators()
            post_partition);
         })
 
-    .op("aten::split(str self, str separator=' ', int max=-1) -> str[]",
+    // Due to type constraints in TorchScript, we use a sentinel string
+    // "__DO_NOT_USE" as the default separator for split functions.
+    // This denotes splitting by any whitespace, or default behavior when
+    // the separator is not specified
+    .op("aten::split(str self, str separator='__DO_NOT_USE', int max=-1) -> str[]",
```
Comment (driazati) [inline torch/csrc/jit/register_string_ops.cpp]: Can `separator` be type `str?` instead?

File: torch/csrc/jit/register_string_ops.cpp [hunk 4]
```diff
@@ -547,48 +547,94 @@ auto reg_str_ops_2 = torch::RegisterOperators()
...
+                separator = sep;
+              }
+            }
+            pos = min_pos;
+            if (pos != std::string::npos) {
+              count++;
+              if (max >= 0 && count > max) {
+                break;
+              } else {
+                splits.emplace_back(string.substr(prev_pos, pos - prev_pos));
```
Comment (driazati) [inline torch/csrc/jit/register_string_ops.cpp]: For a string like `a\t\tb` won't this lead to `['a', '', 'b']`? Python treats all contiguous whitespace as 1 separator

File: torch/csrc/jit/register_string_ops.cpp [hunk 5]
```diff
@@ -547,48 +547,94 @@ auto reg_str_ops_2 = torch::RegisterOperators()
...
-          while ((pos = string.find(separator, pos)) != std::string::npos) {
-            count++;
-            if (max >= 0 && count > max) {
-              break;
+          std::vector<std::string> separators = {separator};
+          if (separator == default_separator) {
+            separators = { "\n", "\t", " " };
+          }
+          do {
+            auto min_pos = std::string::npos;
```
Comment (driazati) [inline torch/csrc/jit/register_string_ops.cpp]: there's a lot of code copy paste between here and rsplit, could they be combined?

File: test/test_jit_string.py (hunk 2 above)
Comment (bwasti) [reply test/test_jit_string.py]: it won't pass the tests on my machine without it -- the default seems to conservative

File: torch/csrc/jit/register_string_ops.cpp (hunk 3 above)
Comment (bwasti) [reply torch/csrc/jit/register_string_ops.cpp]: ooh, thanks! didn't realize this existed -- cleans up a lot

File: torch/csrc/jit/register_string_ops.cpp (hunk 4 above)
Comment (bwasti) [reply torch/csrc/jit/register_string_ops.cpp]: ah yea, nice catch! this is true if the separator isn't specified (and not true if it is)

File: test/test_jit_string.py [hunk 6]
```diff
@@ -1,5 +1,6 @@
 from test_jit import JitTestCase
```
Comment (suo) [inline test/test_jit_string.py]: these tests are not run in oss CI. You need to whitelist them in `run_tests.py`.

File: test/test_jit_string.py (hunk 2 above)
Comment (suo) [reply test/test_jit_string.py]: what is causing the big recursion? Is it not suspicious?

**Review by suo — CHANGES_REQUESTED** [review]

File: test/test_jit_string.py [hunk 7]
```diff
@@ -278,18 +279,20 @@ def test_split():
...
+                "d d d d d ".rsplit(" ", 10),
+                "e e e e e ".rsplit(" ", -1),
+                "f f f f f ".rsplit(" ", 3),
+                " g g g g g ".rsplit("*"),
+                " h*h h*h h ".rsplit("*"),
+                " i*i i*i i ".rsplit("*", -1),
+                " j*j j*j j".rsplit("j*", 10),
             )
+
         self.checkScript(test_rsplit, ())
```
Comment (suo) [inline test/test_jit_string.py]: can you also add an `if __name__ == __main__`  thing at the bottom so we can run this test in isolation?

**Comment by suo** [comment]
@bwasti any update on this?

**Review by suo — APPROVED** [review]
lgtm except there are test failures looks like

**Comment by bwasti** [comment]
this PR is now covered by #25422
